src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala

```diff
@@ -144,21 +149,21 @@ object JsonParser {
       List[Decoder[JsonExpression]](
         Decoder[JsonType].widen,
         Decoder[JsonIntConstant].widen,
-        Decoder[JsonStringConstant].widen,
-        Decoder[JsonEmptyExpression].widen
+        Decoder[JsonStringConstant].widen
+      ).reduceLeft(_ or _)
+
+    val decodeInstructionSequence: Decoder[JsonInstructionSequence] = _.as[List[JsonInstruction]].map(JsonInstructionSequence)
+
+    implicit val decodeInstruction: Decoder[JsonInstruction] =
+      List[Decoder[JsonInstruction]](
+        decodeInstructionSequence.widen,
+        Decoder[JsonSimpleInstruction].widen
       ).reduceLeft(_ or _)
 
-    implicit val decodeInstruction: Decoder[JsonInstruction] = cursor => {
-      lazy val isSequence = (_: HCursor).downArray.succeeded
-      lazy val isEmptyArray = (_: Json).asArray.exists(_.isEmpty)
-      lazy val isComplexInstruction = (_: HCursor).downField("args").downArray.find(it => isSequence(it.hcursor) || isEmptyArray(it)).succeeded
-
-      if (isSequence(cursor))
-        cursor.as[List[JsonInstruction]].map(JsonInstructionSequence)
-      else if (isComplexInstruction(cursor))
-        cursor.as[JsonComplexInstruction]
-      else
-        cursor.as[JsonSimpleInstruction]
+    implicit def decodeEither[A,B](implicit a: Decoder[A], b: Decoder[B]): Decoder[Either[A,B]] = {
```

Comment:
oh my, there's nothing already available in circe to do this? it's surprising

Reply:
The default encoder looking for ``left`` and ``right`` keys.

Reply:
I see